README.md
# cart_counter

A Flutter Package which Helps you can Implement Cart Counter in your flutter project. Its really Simple and Easy to use. This Package will provide me a default design and if you want to change according to you then you can also do this. In This Package you can enable TextField for custom Enter value. simple in CartCounter widget you need to pass isEnable paramter true like: isEnable:true then you'll able to Enter custom value.

## Screenshots


<td>
<img src="https://github.com/chandan123-pradhan/cart_counter/blob/main/cart_counter.jpg" alt="">
</tr>

## Usage
[Example](https://github.com/chandan123-pradhan/cart_counter/blob/main/example/main.dart)

To use this package:

* add the dependency to your [pubspec.yaml](https://github.com/chandan123-pradhan/cart_counter/blob/main/pubspec.yaml) file.

```yaml
    dependencies:
      flutter:
        sdk: flutter
      cart_counter: ^0.0.1
```

### How to use

```dart
  
import 'package:cart_counter/Widgets/CartCounter.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        debugShowCheckedModeBanner: false,
        title: 'Cart Counter',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: HomePage());
  }
}

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Container(
      height: double.infinity,
      child: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
           Container(
              width: 150,
              child: CartCounter(
                maximumValue: 10,
                minimumValue: 3,
                value: 3,
                style: TextStyle(
                    color: Colors.red,
                    fontSize: 30,
                    fontWeight: FontWeight.bold),
               onChanged: (val) {
                  //here in val variable you'll get updated counter value.
                  print(val);
                },
              ),
            ),
          ],
        ),
      ),
    ));
  }
}


```

## Done Thank You For The Using This Package. It Will Be Really Really Helpfull.


CHANGELOG.md
## [0.0.1] - Release

* Implement Simple Decorated Cart Counter Easily.
* You Can Customize This Widget by adding Decoration.
* You Can Enable Text Field in this widget and can Enter Custom value by passing isEnable:true.
* You Can Customize Increament And Decreament Widget.
* You Can Customize Counter Style. by passing style:TextStyle().

lib/Widgets/cart_counter.dart
import 'package:flutter/material.dart';
import 'package:flutter_bounce/flutter_bounce.dart';

class CartCounter extends StatefulWidget {
  /**
   * Minimum Value at Last User can Decrease.
   */
  int minimumValue;
  /**
   * Maximum Value at Last User can Increase.
   */
  int maximumValue;
  /**
   * Current Value.
   */
  int value;
  /**
   * Return updated value.
   */
  Function onChanged;
  /**
   * For Decoration of your main widget.
   */
  BoxDecoration decoration;
  /**
   * TextStyle of Counter.
   */
  TextStyle style;
  /**
   * For Enable Custom Enter value.
   */
  bool isEnable;

  /**
   * for Increase Botton Widget.
   */
  Widget increaseBottonWidget;
  /**
   * for Decrease Botton Widget.
   */
  Widget decreaseBottonWidget;
  CartCounter(
      {@required this.minimumValue,
      @required this.maximumValue,
      @required this.value,
      @required this.onChanged,
      this.decoration,
      @required this.style,
      this.isEnable,
      this.increaseBottonWidget,
      this.decreaseBottonWidget,
     });

  @override
  _CartCounterState createState() => _CartCounterState(
      minimumValue,
      maximumValue,
      value,
      onChanged,
      decoration,
      style,
      isEnable,
      increaseBottonWidget,
      decreaseBottonWidget,
    );
}

class _CartCounterState extends State<CartCounter> {
  /**
   * Minimum Value at Last User can Decrease.
   */
  int minimumValue;
  /**
   * Maximum Value at Last User can Increase.
   */
  int maximumValue;
  /**
   * Current Value.
   */
  int value;
  /**
   * Return updated value.
   */
  Function onChanged;
  
  /**
   * For Decoration of your main widget.
   */
  BoxDecoration decoration;
  /**
   * TextStyle of Counter.
   */
  TextStyle style;
  /**
   * For Enable Custom Enter value.
   */
  bool isEnable;
  /**
   * for Increase Botton Widget.
   */
  Widget increaseBottonWidget;
  /**
   * for Decrease Botton Widget.
   */
  Widget decreaseBottonWidget;
  _CartCounterState(
      this.minimumValue,
      this.maximumValue,
      this.value,
      this.onChanged,
      this.decoration,
      this.style,
      this.isEnable,
      this.increaseBottonWidget,
      this.decreaseBottonWidget,
     );

  TextEditingController valueController = new TextEditingController();

  @override
  void initState() {
    valueController.text = value.toString();
    super.initState();
  }

/**
 * This function will be increase value of counter.
 */
  void _increase() {
    if (value < maximumValue) {
      // This Block will be true when current counter value will be less than maximum value.
      setState(() {
        value++; //increasing counter value.
      });
      valueController.text =
          value.toString(); //assigning updated coutner value to the text value.
      onChanged(value); // calling [onChanged] functions.
    }
  }

/**
 * This function will be decrease value of counter.
 */
  void _decrease() {
    if (value > minimumValue) {
      // This Block will be true when current counter value will be greater than minimum value.
      setState(() {
        value--; //Decreasing counter value.
      });
      valueController.text =
          value.toString(); //assigning updated coutner value to the text value.
      onChanged(value); // calling [onChanged] functions.
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: decoration??BoxDecoration(
          boxShadow: 
              [
                BoxShadow(
                  color: Colors.grey.withOpacity(0.2),
                  spreadRadius: 0.5,
                  blurRadius: 2,
                  offset: Offset(0, 3), // changes position of shadow
                ),
              ],
          borderRadius:BorderRadius.circular(5),
          color: Colors.white),
      child: Padding(
        padding: const EdgeInsets.only(left: 5, right: 5),
        child: Row(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            Bounce(
              duration: Duration(milliseconds: 110),
              onPressed: () {
                _increase(); //calling _increase() function when user will be tap on - icon.
              },
              child: increaseBottonWidget ??
                  Container(
                      decoration: BoxDecoration(
                        shape: BoxShape.circle,
                        color: Colors.red,
                        boxShadow: [
                          BoxShadow(
                            color: Colors.grey.withOpacity(0.5),
                            spreadRadius: 1,
                            blurRadius: 7,
                            offset: Offset(0, 3), // changes position of shadow
                          ),
                        ],
                      ),
                      child: Padding(
                        padding: const EdgeInsets.all(3),
                        child: Icon(
                          Icons.add,
                          size: 30,
                          color: Colors.white,
                        ),
                      )),
            ),
            Flexible(
              child: Container(
                alignment: Alignment.center,
                child: TextField(
                  keyboardType: TextInputType.number,
                    onEditingComplete: () {
                      if (valueController.text.isEmpty) {
                        setState(() {
                          value = minimumValue;
                          valueController.text = minimumValue.toString();
                          onChanged(value);
                        });
                      }
                    },
                    enabled: isEnable,
                    controller: valueController,
                    autofocus: false,
                    textAlign: TextAlign.center,
                    decoration: InputDecoration(
                      border: InputBorder.none,
                      focusedBorder: InputBorder.none,
                      enabledBorder: InputBorder.none,
                      errorBorder: InputBorder.none,
                      disabledBorder: InputBorder.none,
                      //  contentPadding: EdgeInsets.only(left: 15, bottom: 11, top: 11, right: 15),
                    ),
                    style: style ?? //Applying given textStyle.
                        TextStyle(
                          color: Colors.black,
                          fontSize: 18,
                        ),
                    onChanged: (val) {
                    if (val.isNotEmpty) {
                      //if Entered value on TextField will be Not Empty.
                      if (int.parse(val) < maximumValue) {
                        //if Entered Value will be not Greater than Maximum Value.
                        setState(() {
                          value = int.parse(
                              val); //Assigning Entered Value to the Counter Value.
                          valueController.text = value
                              .toString(); //assigning updated coutner value to the text value.
                        });
                        onChanged(
                            value); //Calling onChanged() function for get Updated Counter Value.
                      } else {
                        //if Entered Value will be more than maximum value then Maximum value will be assign.
                        setState(() {
                          value =
                              maximumValue; //Assigning Maximum value to the counter.
                          valueController.text = value
                              .toString(); //Assigning counter updated value to the value controller.
                        });
                        onChanged(
                            value); //calling onChanged() function for get updated counter value.
                      }
                    } else {
                      setState(() {
                        valueController.text = '';
                        //Assigning counter updated value to the value controller.
                      });
                      onChanged(value);
                    }
                  }),
              ),
            ),
            Bounce(
              duration: Duration(milliseconds: 110),
              onPressed: () {
                _decrease(); //calling _decrease() function when user will be tap on - icon.
              },
              child: decreaseBottonWidget ??
                  Container(
                      decoration: BoxDecoration(
                        shape: BoxShape.circle,
                        color: Colors.red,
                        boxShadow: [
                          BoxShadow(
                            color: Colors.grey.withOpacity(0.5),
                            spreadRadius: 1,
                            blurRadius: 7,
                            offset: Offset(0, 3), // changes position of shadow
                          ),
                        ],
                      ),
                      child: Icon(
                        Icons.remove,
                        size: 30,
                        color: Colors.white,
                      )),
            ),
          ],
        ),
      ),
    );
  }
}


example/main.dart

import 'package:cart_counter/Widgets/cart_counter.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        debugShowCheckedModeBanner: false,
        title: 'Cart Counter',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: HomePage());
  }
}

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Container(
      height: double.infinity,
      child: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            // SizedBox(height: 200),
            Text("Cart Counter",
                style: TextStyle(color: Colors.black, fontSize: 20)),
            SizedBox(height: 40),
            Container(
              width: 150,
              child: CartCounter(
                
                increaseBottonWidget: Icon(Icons.add,color:Colors.white,size:40),// if you want to add custom add botton then here you can pass your custom widget.
                decreaseBottonWidget:Icon(Icons.remove,color:Colors.white,size:40),// if you want to add custom Remove botton then here you can pass your custom widget.
               
               // If you want to decorate custom widget here you can pass your body decoration. its optional.
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(20),
                  color: Colors.blue,
                  boxShadow: [
                    BoxShadow(
                      color: Colors.grey.withOpacity(0.2),
                      spreadRadius: 0.5,
                      blurRadius: 2,
                      offset: Offset(0, 3), // changes position of shadow
                    ),
                  ],
                ),
                maximumValue: 10,
                minimumValue: 3,
                value: 3,
                style: TextStyle(
                    color: Colors.white,
                    fontSize: 30,
                    fontWeight: FontWeight.bold),
                isEnable: true,//if you want to Enter custom value then here you should be enable.
                onChanged: (val) {
                  //here in val variable you'll get updated counter value.
                  print(val);
                },
              ),
            ),
          ],
        ),
      ),
    ));
  }
}
